src/core/algorithms/nar/des/encoded_value_range.h

```diff
@@ -0,0 +1,112 @@
+#pragma once
+
+#include "algorithms/nar/value_range.h"
+#include "model/table/column_layout_typed_relation_data.h"
+#include "rng.h"
+
+namespace algos::des {
+
+class EncodedValueRange {
+public:
+    static size_t const kFieldCount = 4;
+    double permutation;
+    double threshold;
+    double bound1;
+    double bound2;
+
+    // TODO: remove code duplication here
+    double& operator[](size_t index) {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    double const& operator[](size_t index) const {
+        switch (index) {
+            case 0:
+                return permutation;
+                break;
+            case 1:
+                return threshold;
+                break;
+            case 2:
+                return bound1;
+                break;
+            case 3:
+                return bound2;
+                break;
+            default:
+                throw std::out_of_range("Index out of range for value range.");
+        }
+    }
+
+    std::shared_ptr<model::ValueRange> Decode(std::shared_ptr<model::ValueRange> domain) const {
+        switch (domain->GetTypeId()) {
+            case model::TypeId::kInt: {
+                std::shared_ptr<model::IntValueRange> int_domain =
+                        std::static_pointer_cast<model::IntValueRange>(domain);
```

Comment:
You can use `auto` when specifying type of a variable when you are performing a cast. The reason is that the type can be easily deduced by both a compiler and a person reading the code:
```suggestion
                auto int_domain = std::static_pointer_cast<model::IntValueRange>(domain);
```